src/main/java/org/springframework/integration/dsl/StandardIntegrationFlow.java

```diff
@@ -16,18 +16,50 @@
 
 package org.springframework.integration.dsl;
 
+import java.util.HashSet;
 import java.util.LinkedHashSet;
 import java.util.Set;
 
+import org.springframework.beans.factory.BeanNameAware;
+import org.springframework.context.Lifecycle;
+import org.springframework.context.SmartLifecycle;
+import org.springframework.integration.dsl.core.EndpointSpec;
+import org.springframework.integration.endpoint.AbstractEndpoint;
+
 /**
 * @author Artem Bilan
 */
-public class StandardIntegrationFlow implements IntegrationFlow {
+public class StandardIntegrationFlow implements IntegrationFlow, SmartLifecycle {
 
 	private final Set<Object> integrationComponents;
 
+	private final Set<Lifecycle> lifecycles = new HashSet<Lifecycle>();
```

Comment:
Perhaps use a `LinkedHashSet` (or a simple `List`) and start last-to-first, stop first-to-last ?

This would avoid `Dispatcher has no subscribers` for in-flight messages.

Similar logic is used in XD for streams.